base/applications/cmdutils/more/more.c

```diff
@@ -121,11 +285,69 @@ PagePrompt(PCON_PAGER Pager, DWORD Done, DWORD Total)
 
         /* Got our key, return to caller */
         KeyEvent = ir.Event.KeyEvent;
-    }
-    while ((KeyEvent.wVirtualKeyCode == VK_SHIFT) ||
-           (KeyEvent.wVirtualKeyCode == VK_MENU) ||
-           (KeyEvent.wVirtualKeyCode == VK_CONTROL));
 
+        /* Ctrl key is pressed? */
+        fCtrl = !!(KeyEvent.dwControlKeyState & (LEFT_CTRL_PRESSED | RIGHT_CTRL_PRESSED));
+
+        /* Ctrl+C or Ctrl+Esc? */
+        if (fCtrl && ((KeyEvent.wVirtualKeyCode == VK_ESCAPE) ||
+                      (KeyEvent.wVirtualKeyCode == L'C')))
+        {
+            s_chSubCommand = 0;
+            break;
+        }
+
+        /* Ignore any unsupported keyboard press */
+        if ((KeyEvent.wVirtualKeyCode == VK_SHIFT) ||
+            (KeyEvent.wVirtualKeyCode == VK_MENU)  ||
+            (KeyEvent.wVirtualKeyCode == VK_CONTROL))
+        {
+            continue;
+        }
+
+        /* Ignore any unsupported subcommand */
+        if (s_chSubCommand != L'P' && s_chSubCommand != L'S')
+            break;
```

Comment:
Technically if your code runs fine, the `s_chSubCommand` variable must only take the values '0' (NULL character), 'P' or 'S' and nothing more, so that might just need to become some sort of assertion on the value of `s_chSubCommand` (see also comments below).

Reply:
Insert an assertion. OK.